### ISSUE TITLE
`unused-parameter` marks unused parameter as used due to identical field in composite literal struct

### ISSUE THREAD

**Issue by mgnsk** [mgechev/revive#721]
**Describe the bug**
The `unused-parameter` rule doesn't detect unused parameter when the function body constructs a struct containing a field with the same name as the parameter.

**To Reproduce**
1. Installed revive from git
```shell
> revive --version
Version:        v1.2.1-22-gdb56db0-dev
Commit:         db56db0b6a11404963d53472d5a40d90cc0868a6
Built           2022-07-29 20:53 UTC by Makefile
```
Config

```toml
# config file
ignoreGeneratedHeader = false
severity = "warning"
confidence = 0.8
errorCode = 1
warningCode = 1

[rule.unused-parameter]
```
2. The following file should give an error

```go
package main

type response struct {
	unused string
}

func example(unused string) {
	_ = response{
		unused: "",
	}
}
```

```shell
> revive -config=revive.toml . # exits with 0
```

**Expected behavior**
Expected an error saying that the `unused` parameter is unused

**Desktop (please complete the following information):**
 - OS: Arch Linux
 - Version of Go: go version go1.18.4 linux/amd64

**Additional context**
Seems like the problems stems from here:
https://github.com/mgechev/revive/blob/db56db0b6a11404963d53472d5a40d90cc0868a6/rule/unused-param.go#L57-L60
As far as I can tell, the selection works on the lexical ident values and there is not enough information to tell in which context the ident is being used.

There's also talk of deprecating `*ast.Object` (https://github.com/golang/go/issues/52463) and using the new `go/types` but I don't think it applies here since resolving types is not needed. The checker should inspect for an `*ast.KeyValueExpr` inside a composite literal whose `Key` matches the parameter ident and then skip that from being counted as a used parameter.

If no-one has any immediate ideas for this, I can start looking into it.

**Comment by mgnsk** [comment]
It's a bit more complicated than that.

Given the following new testcases, `f10` and `f11` fail to match an error because it doesn't detect that `unused` is a struct field. `f12` correctly succeeds.

```go
func f8(unused string) { // MATCH /parameter 'unused' seems to be unused, consider removing or renaming it as _/
	_ = struct {
		unused string
	}{}
}

type responseStruct struct {
	unused string
}

func f9(unused string) { // MATCH /parameter 'unused' seems to be unused, consider removing or renaming it as _/
	r := responseStruct{}
	r.unused = ""
}

func f10(unused string) { // MATCH /parameter 'unused' seems to be unused, consider removing or renaming it as _/
	_ = responseStruct{
		unused: "",
	}
}

func f11(unused string) { // MATCH /parameter 'unused' seems to be unused, consider removing or renaming it as _/
	_ = struct {
		unused string
	}{
		unused: "",
	}
}

type responseMap map[string]interface{}

func f12(unused string) {
	_ = responseMap{
		unused: nil,
	}
}
```

There are some nuances - if the struct was defined in another package, then it only has exported fields. Function parameters usually begin with a lowercase character, this means the struct type is most likely defined in the same package, it should be trivial to detect that the composite literal is a struct type without parsing other packages by looking at `*lint.File`. This has a tradeoff that it doesn't handle imported structs which would imply function parameters beginning with an uppercase character which is unlikely anyway.

```go
// Apply applies the rule to given file.
func (*UnusedParamRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
	...
	for _, f := range file.Pkg.Files() {
		// inspect f for *ast.StructType and collect fields
	}
	...
}
```
It should collect the fields from struct types in the package, then during lint, skip any references that appear as struct fields (we can now distinguish a struct type from a map) from being marked as used and also handle anonymous struct types.

**Comment by chavacava** [comment]
Hi @mgnsk, thanks for filling the issue, the complete analysis and the PR!

As you mentioned, the current implementation uses `ast.Object` to distinguish among all the identifiers and in some cases it is not enough to correctly differentiate two identifiers with the same name (for example a parameter from a struct field)

Leveraging type information solves the problem but at the price of requiring type info calculation. Then the question is: does the amount of false positives justify incrementing complexity and execution time?

**Comment by mgnsk** [comment]
I see now, there's a pretty significant performance drop:
```
BenchmarkUnusedParam-8              1213            957475 ns/op
```
->
```
BenchmarkUnusedParam-8               172           7104002 ns/op
```

The specific case I had was during refactoring old code to add context support, actually using the new `ctx` parameter was forgotten:
```go
func newProcess(ctx context.Context) *process {
    return &process{ctx: context.Background()}
}
```

**Comment by chavacava** [comment]
@mgnsk I'll study the case and see if it could be detected without using type info
